File: packages/server/cmd/renewlet/telegram_format_test.go
package main

import (
	"strings"
	"testing"
)

func TestTelegramFormatterDefaultsToPlainText(t *testing.T) {
	message := notificationMessage{
		Title:     "Renewlet & Friends",
		Content:   "A&B <Plan>",
		Timestamp: "2026-06-20 08:00:00 UTC",
	}

	formatted := formatTelegramNotificationMessage(message, telegramMessageFormatPlain)

	if formatted.ParseMode != "" {
		t.Fatalf("plain notification must not set parse mode, got %q", formatted.ParseMode)
	}
	if formatted.Text != buildTextMessage(message) {
		t.Fatalf("plain notification text drifted: %q", formatted.Text)
	}
}

func TestTelegramFormatterEscapesHTMLMode(t *testing.T) {
	notification := formatTelegramNotificationMessage(notificationMessage{
		Title:     `Renewlet & <Friends>`,
		Content:   `A&B <Plan> "Quote"`,
		Timestamp: "2026-06-20 08:00:00 UTC",
	}, telegramMessageFormatHTML)
	if notification.ParseMode != "HTML" {
		t.Fatalf("expected HTML parse mode, got %#v", notification)
	}
	if !strings.Contains(notification.Text, `<b>Renewlet &amp; &lt;Friends&gt;</b>`) || strings.Contains(notification.Text, "<Friends>") {
		t.Fatalf("notification html was not escaped: %q", notification.Text)
	}

	bot := formatTelegramBotMessage("Renewlet status\nTotal: 12\n总数：12\nA&B <Pro>", telegramMessageFormatHTML)
	if bot.ParseMode != "HTML" {
		t.Fatalf("expected bot HTML parse mode, got %#v", bot)
	}
	if !strings.Contains(bot.Text, "<b>Renewlet status</b>") || !strings.Contains(bot.Text, "Total: <b>12</b>") || !strings.Contains(bot.Text, "总数：<b>12</b>") {
		t.Fatalf("bot html emphasis missing: %q", bot.Text)
	}
	if !strings.Contains(bot.Text, "A&amp;B &lt;Pro&gt;") || strings.Contains(bot.Text, "<Pro>") {
		t.Fatalf("bot html user content was not escaped: %q", bot.Text)
	}
}

File: packages/cloudflare/src/telegram-format.test.ts
import { describe, expect, it } from "vitest";
import { plainNotificationMessage, telegramBotMessage, telegramNotificationMessage } from "./telegram-format";

describe("Telegram message formatter", () => {
  it("keeps plain text as the default notification shape", () => {
    const message = {
      title: "Renewlet & Friends",
      content: "A&B <Plan>",
      timestamp: "2026-06-20 08:00:00 UTC",
      items: [],
      hasPayload: true,
    };

    const formatted = telegramNotificationMessage(message, "plain");

    expect(formatted).toEqual({ text: plainNotificationMessage(message) });
  });

  it("escapes notification and bot content before applying HTML emphasis", () => {
    const notification = telegramNotificationMessage({
      title: "Renewlet & <Friends>",
      content: `A&B <Plan> "Quote"`,
      timestamp: "2026-06-20 08:00:00 UTC",
      items: [],
      hasPayload: true,
    }, "html");
    expect(notification).toMatchObject({ parse_mode: "HTML" });
    expect(notification.text).toContain("<b>Renewlet &amp; &lt;Friends&gt;</b>");
    expect(notification.text).toContain("A&amp;B &lt;Plan&gt; &quot;Quote&quot;");
    expect(notification.text).not.toContain("<Friends>");

    const bot = telegramBotMessage("Renewlet status\nTotal: 12\n总数：12\nA&B <Pro>", "html");
    expect(bot).toMatchObject({ parse_mode: "HTML" });
    expect(bot.text).toContain("<b>Renewlet status</b>");
    expect(bot.text).toContain("Total: <b>12</b>");
    expect(bot.text).toContain("总数：<b>12</b>");
    expect(bot.text).toContain("A&amp;B &lt;Pro&gt;");
    expect(bot.text).not.toContain("<Pro>");
  });
});
